File: 0101-0200/0153.Find_Minimum_in_Rotated_Sorted_Array.cpp
//0ms





class Solution {
public:
    int findMin(vector<int>& nums) {
        int low = 0, high = nums.size() - 1, mini = INT_MAX;
        while (low <= high) {
            int mid = (low + high) / 2;

            // Left half is sorted
            if (nums[low] <= nums[mid]) {
                mini = min(mini, nums[low]);
                low = mid + 1;
            }
            // Right half is sorted
            else {
                mini = min(mini, nums[mid]);
                high = mid - 1;
            }
        }
        return mini;
    }
};






//0ms






class Solution {
public:
    int findMin(vector<int>& nums) {
        int left = 0, right = nums.size() - 1;
        while(left < right){
            int mid = left + (right - left) / 2;
            if(nums[mid] < nums[right]){
                right = mid;
            }
            else {
                left = mid + 1;
            }
        }
        return nums[left];
    }
};
//2:04

File: 0101-0200/0153.Find_Minimum_in_Rotated_Sorted_Array.rs
//0ms




use std::cmp::Ordering::{Greater, Less, Equal};

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let k = Self::find_peak(&nums);
        nums[k%n]
    }

    fn find_peak(nums: &Vec<i32>) -> usize {
        let (mut lbound, mut ubound) = (0, nums.len());
        while lbound < ubound {
            let mid = (lbound+ubound)/2;
            match nums[mid].cmp(&nums[0]) {
                Greater|Equal => {lbound = mid+1;}
                Less => {ubound = mid;}
            }
        }
        ubound
    }
}




//0ms






impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let (mut left, mut right) = (0i32, nums.len() as i32 - 1);

        while left < right {
            let mut mid = left + (right - left) / 2;

            if nums[mid as usize] > nums[right as usize] {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        nums[left as usize]
    }
}
